modules/programs/autorandr.nix

```diff
@@ -114,6 +114,20 @@ let
           for the documentation of the transform matrix.
         '';
       };
+
+      dpi = mkOption {
+        type = types.nullOr types.int;
+        description = "Output DPI configuration.";
+        default = null;
+        example = 96;
+      };
+
+      scale = mkOption {
```

Comment:
use [transform](
https://github.com/rycee/home-manager/blob/d009aeb1758b65c79233200e5bb82f223481f6ed/modules/programs/autorandr.nix#L98) for scaling.

Reply:
I saw that transform was available, but what I gather from the documentation is that `scale` is just an alias around `transform` and I found the syntax more friendly. Is there any reason not to support the different options? Thanks for the input.

Reply:
@kalbasit according to the xrandr [man page](https://www.x.org/releases/X11R7.5/doc/man/man1/xrandr.1.html), `scale` is 

> a shortcut version of the --transform option. 

So I guess it still makes sense to support it for simple cases? We could also implement `scale` option through `transform` by creating correct matrix from the specified values. I'm not very familiar with `transform`, `scale`, and `scale-from` options of xrandr, so correct me if I'm wrong.

@terlar I think it's better to turn scale option into a submodule so that `x` and `y` will be typed and can be set separately, e.g.:

```
scale = mkOption {
  type = types.nullOr (types.submodule {
    options = {
      x = mkOption {
        type = types.float;
        description = "Horizontal scaling factor."
      };
      y = mkOption {
        type = types.float;
        description = "Vertical scaling factor."
      };
    };
  });
  default = null;
  example = { x = 1.25; y = 1.25 };
  description = "Output scale configuration.";
}
```

Reply:
Good ideas, I was keeping it consistent with `mode` and `rate` and seemed this was the approach. But that opens for more improvements, perhaps a PR to type those could be made as well later.

Reply:
I've updated it with the typed definition for `scale` as suggested. If this gets merged, I will make another one with typing of the other things.

Reply:
I don't mind adding another option, especially that it's supported by xrandr. It would useful to highlight that in the documentation as well. Also, are `scale` and `transform` mutually exclusive? Should you add an assertion for that?

Reply:
Using some manual executions to `xrandr` with the different options I can verify that they are mutually exclusive with an interesting twist. Specifying both `scale` and `transform`, whichever is specified last will apply. However specifying `scale-from` in any position will have precedence over both `scale` and `transform`. However that doesn't really matter, we should probably ensure you only have one of these. What is the pattern to add such an assertion, which one should have highest precedence, or would the configuration just fail if you have multiple specified? I am quite new to specifying types inside nix expressions.

Reply:
Would it make sense to combine `scale` and `scale-from` as `scale` and base the target on the type?

If `x = 1.5; y = 1.5;` then it would use `scale` but if the number is an integer e.g. `x = 1920; y = 1080` then it would be interpreted as `scale-from`. The only caveat with this is if someone would expect `x = 1; y = 1;` to scale by factor, however right now we force float anyways.

Reply:
Or if it should be an enum option, such as `method = "pixel"` and `method = "factor"`.

Reply:
@terlar You can use
https://github.com/rycee/home-manager/blob/799a90ecfaa717bc617ffad19383d1cddd4c99ad/modules/programs/rofi.nix#L298-L305
as a model. I would suggest to avoid issues of precedence by simply failing if both are provided, like in the rofi example.

I think having the `method` option is clearest. I suspect that it will lead to confusion if we attempt to determine the type of operation by whether the user gave a int or float. And then we would also have to worry about the case where a user gave `x` as a float and `y` as an int 🙂

Reply:
Thank you very much, I am just not sure how to apply the assertion for the nested "collection" cases, as what I gather from looking at other examples assertion can only be in this context. I guess I have to figure out some way to map these values to an assertion?

Reply:
I got some time now over the weekend, so I think I figured it out. But please have a look at the new code to see if it makes sense. :bowing_man:

Reply:
Yep, I think it looks great! I'll mark this as resolved 🙂